### PR TITLE
Limit number of visible incompatible extensions in sidebar notice

### LINKED ISSUE #11557: Limit the number of visible incompatible extensions in the sidebar notice
## Describe the bug

During the Woo Core Product Town Hall on November 1, 2023, @beaulebens raised the question of limiting the number of visible incompatible extensions in the sidebar notice.  

Limiting the number of visible incompatible extensions leads to a cleaner UI, but causes the risk that a merchant does not notice that a certain extension is incompatible.

## To reproduce

1. Install and activate the following helper plugins from [this zip file](https://github.com/woocommerce/woocommerce-blocks/files/13241424/helper-plugins.zip). 
2. Create a test page and add the Cart block to it.
3. In the editor, click on the Cart block.
4. In the sidebar, 15 incompatible extensions should be listed.

## Current behavior

There's no limit regarding the number of incompatible extensions that we display. If a merchant installs 100+ incompatible extensions, then all of them are displayed in the sidebar.

## Expected behavior

The number of incompatible extensions that are visible by default should be limited to 2. Below the incompatible extensions, a link should be placed that says `x more incompatibilities`. Clicking on that link shows the remaining incompatible extensions.

Screenshot 2023-11-22 at 10 58 25 AM

### PR DESCRIPTION
**Pull request by nielslange**
## What

Fixes #11557

This PR limits the display of incompatible extensions to two, with the option to view additional ones in a dropdown menu. Instead of introducing an according component to show remaining incompatible extensions, I decided to use the HTML element `<details>`. 

The proposed design in #11557 shows that the `<details>` element does not have the marker on the left-hand side, but a chevron on the right-hand side. For consistency-reasons, I used the chevron from `@wordpress/icons`, which is used to collapse various sidebar sections. For Safari, I had to introduce a separate definition to remove the marker (see review below). 

In https://github.com/woocommerce/woocommerce-blocks/issues/11557#issuecomment-1823149114, @elizaan36 mentioned that the link color should be changed to gray and be underlined. While cross-browser testing, I noticed that Safari does not allow adding `text-decoration` to the `<summary>` element. Therefore, I introduced a `<span>` to ensure that the link is underlined in all browsers.

## Why

This PR enhances the user experience by creating a more clear UI. It also prevents overwhelming merchants by showing too many incompatible extensions by default.

## Testing Instructions

_Please consider any edge cases this change may have, and also other areas of the product this may impact._

> [!NOTE]
> Things to keey an eye on while testing:
> - Number of visible extensions by default and within dropdown.
> - String for exact 1 vs. more than 1 incompatible extension.
> - Chevron changed direction based on opening/closing the dropdown.
> - Link color is gray and link is underlined.
> - The dropdown functionality works is cross-browser compatible.

<details>
<summary>
Test with 2 extensions
</summary>

1. Install and activate the following helper plugins:
- [helper-plugin-01.zip](https://github.com/woocommerce/woocommerce-blocks/files/13487656/helper-plugin-01.zip)
- [helper-plugin-02.zip](https://github.com/woocommerce/woocommerce-blocks/files/13487657/helper-plugin-02.zip)
2. Create a test page and add the Cart block.
3. Select the Cart block and open the sidebar.
4. Verify that the two incompatible extensions are visible.
5. Verify that no dropdown, to more extensions, is visible.
6. Repeat steps 2. until 5. using the Checkout block.

Screenshot 2023-11-28 at 19 13 46

</details>

<details>
<summary>
Test with 3 extensions
</summary>

1. Install and activate the following helper plugins:
- [helper-plugin-01.zip](https://github.com/woocommerce/woocommerce-blocks/files/13487757/helper-plugin-01.zip)
- [helper-plugin-02.zip](https://github.com/woocommerce/woocommerce-blocks/files/13487758/helper-plugin-02.zip)
- [helper-plugin-03.zip](https://github.com/woocommerce/woocommerce-blocks/files/13487759/helper-plugin-03.zip)
2. Create a test page and add the Cart block.
3. Select the Cart block and open the sidebar.
4. Verify that the two incompatible extensions are visible by default.
5. Verify that the dropdown says `1 more incompatibility`.
6.  Click the dropdown link and verify that 1 incompatible extension is visible.
7. Repeat steps 2. until 6. using the Checkout block.

<table>
<tr>
<td valign="top">Closed:
<br><br>
Screenshot 2023-11-28 at 21 32 30
</td>
<td valign="top">Open:
<br><br>
Screenshot 2023-11-28 at 21 32 38
</td>
</tr>
</table>

</details>

<details>
<summary>
Test with 10 extensions
</summary>

1. Install and activate the following helper plugins:
- [helper-plugin-01.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488042/helper-plugin-01.zip)
- [helper-plugin-02.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488043/helper-plugin-02.zip)
- [helper-plugin-03.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488045/helper-plugin-03.zip)
- [helper-plugin-04.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488046/helper-plugin-04.zip)
- [helper-plugin-05.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488047/helper-plugin-05.zip)
- [helper-plugin-06.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488049/helper-plugin-06.zip)
- [helper-plugin-07.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488050/helper-plugin-07.zip)
- [helper-plugin-08.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488051/helper-plugin-08.zip)
- [helper-plugin-09.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488052/helper-plugin-09.zip)
- [helper-plugin-10.zip](https://github.com/woocommerce/woocommerce-blocks/files/13488053/helper-plugin-10.zip)
2. Create a test page and add the Cart block.
3. Select the Cart block and open the sidebar.
4. Verify that the two incompatible extensions are visible by default.
5. Verify that the dropdown says `8 more incompatibilities`. 
7.  Click the dropdown link and verify that 8 incompatible extension are visible.
8. Repeat steps 2. until 6. using the Checkout block.

<table>
<tr>
<td valign="top">Closed:
<br><br>
Screenshot 2023-11-28 at 21 24 23
</td>
<td valign="top">Open:
<br><br>
Screenshot 2023-11-28 at 21 30 26
</td>
</tr>
</table>

</details>

* [ ] Do not include in the Testing Notes 
* [ ] Should be tested by the development team exclusively 

### Before and after

<table>
<tr>
<td valign="top">Before:
<br><br>
Screenshot 2023-11-28 at 19 52 38

</td>
<td valign="top">After (closed):
<br><br>
Screenshot 2023-11-28 at 21 24 23
</td>
<td valign="top">After (open):
<br><br>
Screenshot 2023-11-28 at 21 30 26
</td>

</tr>
</table>

### Cross-browser compatibility

> [!NOTE]  
> - Left: Chrome
> - Center: Firefox
> - Right: Safari

Screenshot 2023-11-28 at 21 26 46
<br><br>
Screenshot 2023-11-28 at 21 26 59

## WooCommerce Visibility

Required:

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental
* [ ] N/A

## Checklist

Required:

* [x] This PR has either a `[type]` label or a `[skip-changelog]` label.
* [x] This PR is assigned to a milestone.

Conditional:

* [x] This PR has a UI change and has been cross-browser tested at different viewport sizes on both the frontend and in the editor.
* [x] This PR has a changelog description (if `[skip-changelog]` label is not present).
* [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR adds/removes an experimental interfaces, and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
* [ ] This PR has been accessibility tested.
* [ ] This PR has had any necessary documentation added/updated.

## Changelog

> Limit number of visible incompatible extensions in sidebar notice

### CODE REVIEW

File: assets/js/editor-components/incompatible-extension-notice/index.tsx [hunk 1]
```diff
@@ -138,6 +138,9 @@ export function IncompatibleExtensionsNotice( {
 		}
 	};
 
+	const entries = Object.entries( incompatiblePaymentMethods );
```
Comment (opr) [inline assets/js/editor-components/incompatible-extension-notice/index.tsx]: Could we take this opportunity to update the name of this var, `incompatiblePaymentMethods` doesn't represent very well what it actually contains

**Review by opr — CHANGES_REQUESTED** [review]
@nielslange this PR is working as expected, I have made an inline suggestion, and I do have another suggestion about the icon. Maybe moving the chevron down a few pixels when it's expanded would be better, toggling between open/closed it looks like it is not aligned correctly with the text when it's expanded.

File: assets/js/editor-components/incompatible-extension-notice/index.tsx (hunk 1 above)
Comment (nielslange) [reply assets/js/editor-components/incompatible-extension-notice/index.tsx]: Good catch, @opr. I've applied the following refactoring:

- Rename `incompatiblePaymentMethods` to `incompatibleExtensions`
- Rename `numberOfIncompatiblePaymentMethods` to `incompatibleExtensionsCount`

**Comment by nielslange** [comment]
> @nielslange this PR is working as expected, I have made an inline suggestion, and I do have another suggestion about the icon. Maybe moving the chevron down a few pixels when it's expanded would be better, toggling between open/closed it looks like it is not aligned correctly with the text when it's expanded.

Thanks for your review, @opr. Regarding the icon, I tried to mimic the behaviour of Gutenberg. That said, I just applied your suggestion and moved the chevron 2px down when it's opened. It looks indeed better, but it looks slightly different from the behavior of the Gutenberg chevron. I'm looking forward to hearing your thoughts on that.

**Review by opr — CHANGES_REQUESTED** [review]
>  but it looks slightly different from the behavior of the Gutenberg chevron. I'm looking forward to hearing your thoughts on that.

Hmm, OK let's keep it consistent, thanks for trying the change out Niels, revert your commit and I'll approve 🙏🏼

**Comment by nielslange** [comment]
> > but it looks slightly different from the behavior of the Gutenberg chevron. I'm looking forward to hearing your thoughts on that.
> 
> Hmm, OK let's keep it consistent, thanks for trying the change out Niels, revert your commit and I'll approve 🙏🏼

Thanks, @opr. I just reverted the chevron commit and all tests passed.

**Review by opr — APPROVED** [review]
Looks good! Thanks Niels.

**Pull request merged**